File: app/src/main/java/com/example/inmocontrol_v2/hid/HidService.java
package com.example.inmocontrol_v2.hid;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothHidDevice;
import android.bluetooth.BluetoothHidDeviceAppSdpSettings;
import android.bluetooth.BluetoothProfile;
import android.content.Intent;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

public class HidService extends Service {
    public static final String TAG = "InmoHID";
    private BluetoothAdapter adapter;
    private BluetoothHidDevice hid;
    private BluetoothDevice lastHost;

    public class LocalBinder extends Binder {
        public HidService getService() {
            return HidService.this;
        }
    }

    private final IBinder binder = new LocalBinder();

    public interface ClientHook {
        void onReady(HidService s);
    }

    private static ClientHook hook;

    public static void setClientListener(ClientHook h) {
        hook = h;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        adapter = BluetoothAdapter.getDefaultAdapter();
        startForeground(1, buildNotif("Idle"));
        if (adapter != null) {
            adapter.getProfileProxy(this, profListener, BluetoothProfile.HID_DEVICE);
        }
        if (hook != null) hook.onReady(this);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    private final BluetoothProfile.ServiceListener profListener = new BluetoothProfile.ServiceListener() {
        @Override
        public void onServiceConnected(int profile, BluetoothProfile proxy) {
            if (Build.VERSION.SDK_INT >= 28) {
                hid = (BluetoothHidDevice) proxy;
                registerApp();
            }
        }
        @Override
        public void onServiceDisconnected(int profile) {
            hid = null;
        }
    };

    private final BluetoothHidDeviceAppSdpSettings sdp = Build.VERSION.SDK_INT >= 28 ?
            new BluetoothHidDeviceAppSdpSettings(
                    "InmoControl HID",
                    "Wear HID controller",
                    "Inmo",
                    BluetoothHidDevice.SUBCLASS1_COMBO,
                    getReportMap()
            ) : null;

    private final BluetoothHidDevice.Callback cb = Build.VERSION.SDK_INT >= 28 ?
            new BluetoothHidDevice.Callback() {
                @Override
                public void onAppStatusChanged(BluetoothDevice dev, boolean registered) {
                    if (registered) {
                        lastHost = dev;
                        autoConnectIfNeeded();
                    }
                }
                @SuppressLint("MissingPermission")
                @Override
                public void onConnectionStateChanged(BluetoothDevice d, int s) {
                    if (s == BluetoothProfile.STATE_CONNECTED) {
                        lastHost = d;
                        saveLast(d);
                        updateNotif("Connected to " + d.getName());
                    } else if (s == BluetoothProfile.STATE_DISCONNECTED) {
                        updateNotif("Disconnected");
                        autoConnectIfNeeded();
                    }
                }
            } : null;

    private void registerApp() {
        if (hid == null || sdp == null || cb == null) return;
        if (Build.VERSION.SDK_INT < 28 || !hasBluetoothPermissions()) return;
        try {
            hid.registerApp(sdp, null, null, Runnable::run, cb);
        } catch (SecurityException e) {
            Log.e(TAG, "Exception", e);
        }
    }

    private void autoConnectIfNeeded() {
        BluetoothDevice d = loadLast();
        if (d != null && hid != null && Build.VERSION.SDK_INT >= 28 && hasBluetoothPermissions()) {
            try {
                hid.connect(d);
            } catch (SecurityException t) {
                Log.e(TAG, "autoConnect failed", t);
            }
        }
    }

    // Public API
    public void sendText(String t) {
        if (hid == null || lastHost == null || t == null) return;
        for (char c : t.toCharArray()) sendKey(asciiToKey(c));
        releaseKeys();
    }

    public void leftClick() {
        sendMouse((byte) 1, 0, 0, 0);
        sendMouse((byte) 0, 0, 0, 0);
    }

    public void rightClick() {
        sendMouse((byte) 2, 0, 0, 0);
        sendMouse((byte) 0, 0, 0, 0);
    }

    public void move(int dx, int dy) {
        sendMouse((byte) 0, clamp(dx), clamp(dy), 0);
    }

    public void wheel(int w) {
        sendMouse((byte) 0, 0, 0, clamp(w));
    }

    public void dpad(String dir) {
        int[] k;
        switch (dir) {
            case "up": k = new int[]{0x52}; break;
            case "down": k = new int[]{0x51}; break;
            case "left": k = new int[]{0x50}; break;
            case "right": k = new int[]{0x4F}; break;
            case "upleft": k = new int[]{0x52, 0x50}; break;
            case "upright": k = new int[]{0x52, 0x4F}; break;
            case "downleft": k = new int[]{0x51, 0x50}; break;
            case "downright": k = new int[]{0x51, 0x4F}; break;
            default: k = new int[]{0x28};
        }
        sendCombo(k);
        releaseKeys();
    }

    public void consumer(String w) {
        int u = 0;
        switch (w) {
            case "playpause": u = 0xCD; break;
            case "vol+": u = 0xE9; break;
            case "vol-": u = 0xEA; break;
            case "forward": u = 0xB5; break;
            case "rewind": u = 0xB6; break;
            case "mute": u = 0xE2; break;
        }
        sendConsumer(u);
    }

    private void sendMouse(byte buttons, int dx, int dy, int wheel) {
        if (hid == null || lastHost == null || Build.VERSION.SDK_INT < 28 || !hasBluetoothPermissions()) return;
        byte[] r = new byte[]{buttons, (byte) dx, (byte) dy, (byte) wheel};
        try {
            hid.sendReport(lastHost, 2, r);
        } catch (SecurityException e) {
            Log.e(TAG, "Exception", e);
        }
    }

    private void sendCombo(int[] k) {
        if (hid == null || lastHost == null || Build.VERSION.SDK_INT < 28 || !hasBluetoothPermissions()) return;
        byte[] rep = new byte[8];
        for (int i = 0; i < k.length && i < 6; i++) {
            rep[2 + i] = (byte) k[i];
        }
        try {
            hid.sendReport(lastHost, 1, rep);
        } catch (SecurityException e) {
            Log.e(TAG, "Exception", e);
        }
    }

    private void sendKey(int k) {
        sendCombo(new int[]{k});
    }

    private void releaseKeys() {
        if (hid == null || lastHost == null || Build.VERSION.SDK_INT < 28 || !hasBluetoothPermissions()) return;
        byte[] r = new byte[8];
        try {
            hid.sendReport(lastHost, 1, r);
        } catch (SecurityException e) {
            Log.e(TAG, "Exception", e);
        }
    }

    private void sendConsumer(int u) {
        if (hid == null || lastHost == null || u == 0) return;
        if (Build.VERSION.SDK_INT < 28) return;
        if (!hasBluetoothPermissions()) return;
        byte[] r = new byte[]{(byte) (u & 0xFF), (byte) ((u >> 8) & 0xFF)};
        try {
            hid.sendReport(lastHost, 3, r);
            hid.sendReport(lastHost, 3, new byte[]{0, 0});
        } catch (SecurityException e) {
            Log.e(TAG, "Exception", e);
        }
    }

    private boolean hasBluetoothPermissions() {
        return checkSelfPermission("android.permission.BLUETOOTH") == android.content.pm.PackageManager.PERMISSION_GRANTED &&
               checkSelfPermission("android.permission.BLUETOOTH_ADMIN") == android.content.pm.PackageManager.PERMISSION_GRANTED &&
               (Build.VERSION.SDK_INT < 31 || checkSelfPermission("android.permission.BLUETOOTH_CONNECT") == android.content.pm.PackageManager.PERMISSION_GRANTED);
    }

    private static int clamp(int v) {
        return Math.max(-127, Math.min(127, v));
    }

    private static int asciiToKey(char c) {
        if (c >= 'a' && c <= 'z') return 0x04 + (c - 'a');
        if (c >= 'A' && c <= 'Z') return 0x04 + (c - 'A');
        if (c >= '1' && c <= '9') return 0x1E + (c - '1');
        if (c == '0') return 0x27;
        if (c == ' ') return 0x2C;
        if (c == '\n') return 0x28;
        if (c == '.') return 0x37;
        if (c == ',') return 0x36;
        if (c == '-') return 0x2D;
        return 0x2C;
    }

    private void saveLast(BluetoothDevice d) {
        if (d == null) return;
        getSharedPreferences("inmo", MODE_PRIVATE)
                .edit()
                .putString("lastHost", d.getAddress())
                .apply();
    }

    private BluetoothDevice loadLast() {
        String addr = getSharedPreferences("inmo", MODE_PRIVATE).getString("lastHost", null);
        if (addr == null || adapter == null) return null;
        try {
            return adapter.getRemoteDevice(addr);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    private Notification buildNotif(String t) {
        NotificationManager nm = getSystemService(NotificationManager.class);
        String ch = "hid";
        if (nm != null) {
            NotificationChannel channel = new NotificationChannel(ch, "Inmo HID", NotificationManager.IMPORTANCE_MIN);
            nm.createNotificationChannel(channel);
        }
        Notification.Builder b = new Notification.Builder(this, ch);
        return b.setContentTitle("InmoControl")
                .setContentText(t)
                .setSmallIcon(android.R.drawable.stat_sys_data_bluetooth)
                .build();
    }

    private void updateNotif(String t) {
        NotificationManager nm = getSystemService(NotificationManager.class);
        if (nm != null) {
            try {
                if (checkSelfPermission("android.permission.POST_NOTIFICATIONS") == android.content.pm.PackageManager.PERMISSION_GRANTED) {
                    nm.notify(1, buildNotif(t));
                }
            } catch (SecurityException e) {
                Log.e(TAG, "Notification update failed", e);
            }
        }
    }

    private static byte[] getReportMap() {
        return new byte[]{
            // Keyboard (ID 1)
            (byte) 0x05, 0x01, (byte) 0x09, 0x06, (byte) 0xA1, 0x01, (byte) 0x85, 0x01, (byte) 0x05, 0x07,
            (byte) 0x19, 0x00, (byte) 0x29, (byte) 0x73, (byte) 0x15, 0x00, (byte) 0x25, 0x01,
            (byte) 0x75, 0x01, (byte) 0x95, 0x08, (byte) 0x81, 0x02, (byte) 0x95, 0x01, (byte) 0x75, 0x08,
            (byte) 0x81, 0x01, (byte) 0x95, 0x05, (byte) 0x75, 0x01, (byte) 0x05, 0x08, (byte) 0x19, 0x01,
            (byte) 0x29, 0x05, (byte) 0x91, 0x02, (byte) 0x95, 0x01, (byte) 0x75, 0x03, (byte) 0x91, 0x01,
            (byte) 0x95, 0x06, (byte) 0x75, 0x08, (byte) 0x15, 0x00, (byte) 0x25, (byte) 0x73, (byte) 0x05, 0x07,
            (byte) 0x19, 0x00, (byte) 0x29, (byte) 0x73, (byte) 0x81, 0x00, (byte) 0xC0,
            // Mouse (ID 2) with X,Y,Wheel
            (byte) 0x05, 0x01, (byte) 0x09, 0x02, (byte) 0xA1, 0x01, (byte) 0x85, 0x02, (byte) 0x09, 0x01,
            (byte) 0xA1, 0x00, (byte) 0x05, 0x09, (byte) 0x19, 0x01, (byte) 0x29, 0x03, (byte) 0x15, 0x00,
            (byte) 0x25, 0x01, (byte) 0x95, 0x03, (byte) 0x75, 0x01, (byte) 0x81, 0x02, (byte) 0x95, 0x01,
            (byte) 0x75, 0x05, (byte) 0x81, 0x01, (byte) 0x05, 0x01, (byte) 0x09, 0x30, (byte) 0x09, 0x31,
            (byte) 0x09, 0x38, (byte) 0x15, (byte) 0x81, (byte) 0x25, 0x7F, (byte) 0x75, 0x08, (byte) 0x95, 0x03,
            (byte) 0x81, 0x06, (byte) 0xC0, (byte) 0xC0,
            // Consumer (ID 3)
            (byte) 0x05, 0x0C, (byte) 0x09, 0x01, (byte) 0xA1, 0x01, (byte) 0x85, 0x03, (byte) 0x15, 0x00,
            (byte) 0x26, (byte) 0xFF, 0x03, (byte) 0x19, 0x00, (byte) 0x2A, (byte) 0xFF, 0x03, (byte) 0x75, 0x10,
            (byte) 0x95, 0x01, (byte) 0x81, 0x00, (byte) 0xC0
        };
    }
}

File: app/src/main/java/com/example/inmocontrol_v2/hid/HidService.kt
package com.example.inmocontrol_v2.hid

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import android.util.Log
import androidx.core.app.NotificationCompat
import com.example.inmocontrol_v2.bluetooth.HidDeviceApp
import com.example.inmocontrol_v2.bluetooth.HidDeviceProfile
import com.example.inmocontrol_v2.bluetooth.HidInputManager
import kotlinx.coroutines.*

/**
 * Core HID orchestrator - Manages Bluetooth HID profile registration, device connections,
 * and input dispatching.
 *
 * Critical Rules from Copilot Instructions:
 * - Register/unregister HID profile on main thread only
 * - Release proxy resources to prevent memory leaks
 * - Maintain foreground notification while paired (Samsung doze prevention)
 * - Retry registerApp() with exponential backoff on STATE_DISCONNECTING
 */
class HidService : Service(), HidServiceApi {

    companion object {
        private const val TAG = "HidService"
        private const val NOTIFICATION_ID = 1001
        private const val CHANNEL_ID = "hid_service_channel"
    }

    inner class LocalBinder : Binder() {
        fun getService(): HidService = this@HidService
    }

    // Coroutine scope for async operations
    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private lateinit var bluetoothAdapter: BluetoothAdapter
    private lateinit var hidDeviceProfile: HidDeviceProfile
    private lateinit var hidDeviceApp: HidDeviceApp
    private lateinit var hidInputManager: HidInputManager

    private var connectedDevice: BluetoothDevice? = null
    private var isServiceReady = false
    private var isProfileConnected = false
    private var isAppRegistered = false

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        startForeground(NOTIFICATION_ID, createNotification())
        initializeBluetoothComponents()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder = LocalBinder()

    override fun onDestroy() {
        super.onDestroy()
        cleanup()
    }

    private fun initializeBluetoothComponents() {
        serviceScope.launch {
            try {
                val bluetoothManager = getSystemService(BluetoothManager::class.java)
                bluetoothAdapter = bluetoothManager.adapter

                hidDeviceProfile = HidDeviceProfile()
                hidDeviceApp = HidDeviceApp()
                hidInputManager = HidInputManager(hidDeviceProfile)

                hidDeviceApp.setDeviceStateListener(deviceStateListener)
                hidDeviceProfile.registerServiceListener(this@HidService, serviceStateListener)
                hidDeviceProfile.registerDeviceListener(profileDeviceStateListener)

            } catch (e: Exception) {
                Log.e(TAG, "Failed to initialize Bluetooth components", e)
                HidClient.setConnectionError("Bluetooth init failed")
            }
        }
    }

    // Device state listener
    private val deviceStateListener = object : HidDeviceApp.DeviceStateListener {
        override fun onConnectionStateChanged(device: BluetoothDevice, state: Int) {
            serviceScope.launch {
                when (state) {
                    BluetoothProfile.STATE_CONNECTED -> {
                        connectedDevice = device
                        hidInputManager.setConnectedDevice(device)
                        HidClient.setConnectionState(true)
                        HidClient.setConnectionError(null)
                        try {
                            Log.d(TAG, "Device connected: ${device.name ?: device.address}")
                        } catch (_: SecurityException) {
                            Log.d(TAG, "Device connected: ${device.address}")
                        }
                    }
                    BluetoothProfile.STATE_DISCONNECTED -> {
                        connectedDevice = null
                        hidInputManager.setConnectedDevice(null)
                        HidClient.setConnectionState(false)
                        try {
                            Log.d(TAG, "Device disconnected: ${device.name ?: device.address}")
                        } catch (_: SecurityException) {
                            Log.d(TAG, "Device disconnected: ${device.address}")
                        }
                    }
                    BluetoothProfile.STATE_CONNECTING -> {
                        try {
                            Log.d(TAG, "Device connecting: ${device.name ?: device.address}")
                        } catch (_: SecurityException) {
                            Log.d(TAG, "Device connecting: ${device.address}")
                        }
                    }
                    BluetoothProfile.STATE_DISCONNECTING -> {
                        try {
                            Log.d(TAG, "Device disconnecting: ${device.name ?: device.address}")
                        } catch (_: SecurityException) {
                            Log.d(TAG, "Device disconnecting: ${device.address}")
                        }
                        // DO NOT call retryRegisterApp() here - this is a device state change,
                        // not a profile service disconnection. Device disconnecting is normal
                        // during connection handshake and should not trigger re-registration.
                    }
                }
            }
        }

        override fun onAppStatusChanged(registered: Boolean) {
            isAppRegistered = registered
            updateServiceReadyState()
            Log.d(TAG, "App registration status: $registered")
        }
    }

    // Service state listener
    private val serviceStateListener = object : HidDeviceProfile.ServiceStateListener {
        override fun onServiceStateChanged(proxy: BluetoothProfile?) {
            Log.d(TAG, "HID profile service state changed: ${proxy != null}")

            isProfileConnected = proxy != null

            if (proxy != null) {
                // Profile connected - register app and WAIT for confirmation
                Log.d(TAG, "Profile connected, starting app registration...")
                serviceScope.launch {
                    delay(100) // Small delay for stability
                    hidDeviceProfile.registerApp(hidDeviceApp)
                }
                // DO NOT set isServiceReady here - wait for onAppStatusChanged callback
            } else {
                // Profile disconnected - reset all flags
                isProfileConnected = false
                isAppRegistered = false
                Log.d(TAG, "Profile disconnected - resetting service state")
            }

            updateServiceReadyState()
        }
    }

    // Profile device state listener
    private val profileDeviceStateListener = object : HidDeviceProfile.DeviceStateListener {
        override fun onConnectionStateChanged(device: BluetoothDevice, state: Int) {
            if (Log.isLoggable(TAG, Log.DEBUG)) {
                Log.d(TAG, "Profile device state: $state")
            }
        }

        override fun onAppStatusChanged(device: BluetoothDevice, registered: Boolean) {
            if (Log.isLoggable(TAG, Log.DEBUG)) {
                Log.d(TAG, "Profile app status: $registered")
            }
        }
    }

    /**
     * Update service ready state based on profile connection AND app registration.
     * Service is only ready when BOTH conditions are met.
     */
    private fun updateServiceReadyState() {
        val wasReady = isServiceReady

        // CRITICAL: Service is only ready when BOTH conditions are met
        isServiceReady = isProfileConnected && isAppRegistered

        if (isServiceReady != wasReady) {
            Log.d(TAG, "Service ready state changed: $isServiceReady (profile=$isProfileConnected, app=$isAppRegistered)")
            HidClient.setServiceReady(isServiceReady)
            HidClient.setHidProfileAvailable(isServiceReady)

            // If we just became ready, notify HidClient immediately
            if (isServiceReady) {
                HidClient.clearError()
                Log.d(TAG, "✅ HID Service fully initialized and ready for connections")
            }
        }
    }

    /**
     * Retry registerApp() with exponential backoff (100ms → 500ms → 1s)
     */
    private suspend fun retryRegisterApp() {
        val delays = listOf(100L, 500L, 1000L)
        for (delay in delays) {
            delay(delay)
            if (hidDeviceProfile.registerApp(hidDeviceApp)) {
                Log.d(TAG, "Successfully re-registered HID app after ${delay}ms")
                break
            }
        }
    }

    private fun createNotificationChannel() {
        val channel = NotificationChannel(
            CHANNEL_ID,
            "HID Service",
            NotificationManager.IMPORTANCE_HIGH  // Changed from LOW to HIGH to prevent app suspension
        ).apply {
            description = "Keeps Bluetooth HID connection active"
            setShowBadge(false)  // Don't show notification badge
            enableVibration(false)  // Don't vibrate
            enableLights(false)  // Don't use LED
        }

        val notificationManager = getSystemService(NotificationManager::class.java)
        notificationManager.createNotificationChannel(channel)
    }

    private fun createNotification(): Notification {
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("InmoWatch HID Active")
            .setContentText("Bluetooth controller ready")
            .setSmallIcon(android.R.drawable.ic_menu_compass)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_HIGH)  // High priority keeps app alive
            .setCategory(NotificationCompat.CATEGORY_SERVICE)  // Mark as service
            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)  // Immediate start
            .build()
    }

    private fun cleanup() {
        serviceScope.cancel()
        hidInputManager.cleanup()
        hidDeviceProfile.unregisterApp()
        hidDeviceProfile.cleanup(bluetoothAdapter)
        connectedDevice = null
    }

    // ==================== HidServiceApi Implementation ====================

    override fun isReady(): Boolean = isServiceReady && isAppRegistered

    override fun isDeviceConnected(): Boolean = connectedDevice != null

    override fun connectToDevice(device: BluetoothDevice): Boolean {
        // CRITICAL: Strict ready check before attempting connection
        if (!isServiceReady) {
            val reason = when {
                !isProfileConnected -> "HID profile not connected"
                !isAppRegistered -> "HID app not registered"
                else -> "Service not initialized"
            }
            Log.e(TAG, "❌ Cannot connect to ${device.address}: $reason")
            HidClient.setConnectionError("Service not ready: $reason")
            return false
        }

        if (!isProfileConnected || !isAppRegistered) {
            Log.w(TAG, "Service not ready or app not registered; cannot connect")
            HidClient.setConnectionError("Service not fully initialized")
            return false
        }
        
        Log.d(TAG, "✅ Connecting to device: ${device.address}")
        val success = hidDeviceProfile.connect(device)
        if (!success) {
            HidClient.setConnectionError("Failed to initiate connection to ${device.address}")
            Log.e(TAG, "❌ Connection initiation failed")
        }
        return success
    }

    // Mouse operations
    override fun sendMouseMovement(deltaX: Float, deltaY: Float): Boolean {
        return hidInputManager.sendMouseMovement(deltaX, deltaY)
    }

    override fun sendLeftClick(): Boolean {
        return hidInputManager.sendMouseClick(left = true, right = false, middle = false)
    }

    override fun sendRightClick(): Boolean {
        return hidInputManager.sendMouseClick(left = false, right = true, middle = false)
    }

    override fun sendMiddleClick(): Boolean {
        return hidInputManager.sendMouseClick(left = false, right = false, middle = true)
    }

    override fun sendDoubleClick(): Boolean {
        val success1 = sendLeftClick()
        Thread.sleep(50) // Small delay between clicks
        val success2 = sendLeftClick()
        return success1 && success2
    }

    override fun sendMouseScroll(deltaX: Float, deltaY: Float): Boolean {
        return hidInputManager.sendMouseScroll(deltaX, deltaY)
    }

    // Keyboard operations
    override fun sendKey(keyCode: Int, modifiers: Int): Boolean {
        return hidInputManager.sendKey(keyCode, modifiers)
    }

    override fun sendText(text: String): Boolean {
        return hidInputManager.sendText(text)
    }

    // Media controls
    override fun playPause(): Boolean {
        return hidInputManager.playPause()
    }

    override fun nextTrack(): Boolean {
        return hidInputManager.nextTrack()
    }

    override fun previousTrack(): Boolean {
        return hidInputManager.previousTrack()
    }

    override fun volumeUp(): Boolean {
        return hidInputManager.volumeUp()
    }

    override fun volumeDown(): Boolean {
        return hidInputManager.volumeDown()
    }

    override fun mute(): Boolean {
        return hidInputManager.mute()
    }

    // D-Pad navigation
    override fun dpad(direction: Int): Boolean {
        // Map direction to appropriate HID command
        return when (direction) {
            0 -> sendDpadUp()
            1 -> sendDpadDown()
            2 -> sendDpadLeft()
            3 -> sendDpadRight()
            4 -> sendDpadCenter()
            // Diagonals: Send two keys simultaneously
            5 -> sendDpadDownLeft()     // ↙
            6 -> sendDpadDownRight()    // ↘
            7 -> sendDpadUpLeft()       // ↖
            8 -> sendDpadUpRight()      // ↗
            else -> false
        }
    }

    override fun sendDpadUp(): Boolean {
        return hidInputManager.sendKey(0x52) // UP_ARROW
    }

    override fun sendDpadDown(): Boolean {
        return hidInputManager.sendKey(0x51) // DOWN_ARROW
    }

    override fun sendDpadLeft(): Boolean {
        return hidInputManager.sendKey(0x50) // LEFT_ARROW
    }

    override fun sendDpadRight(): Boolean {
        return hidInputManager.sendKey(0x4F) // RIGHT_ARROW
    }

    override fun sendDpadCenter(): Boolean {
        return hidInputManager.sendKey(0x28) // ENTER
    }

    // Diagonal movements - send two keys simultaneously
    private fun sendDpadUpLeft(): Boolean {
        return hidInputManager.sendKeys(intArrayOf(0x52, 0x50)) // UP + LEFT
    }

    private fun sendDpadUpRight(): Boolean {
        return hidInputManager.sendKeys(intArrayOf(0x52, 0x4F)) // UP + RIGHT
    }

    private fun sendDpadDownLeft(): Boolean {
        return hidInputManager.sendKeys(intArrayOf(0x51, 0x50)) // DOWN + LEFT
    }

    private fun sendDpadDownRight(): Boolean {
        return hidInputManager.sendKeys(intArrayOf(0x51, 0x4F)) // DOWN + RIGHT
    }

    // ESC key for back functionality
    override fun sendEscape(): Boolean {
        return hidInputManager.sendKey(0x29) // ESC
    }

    // Raw HID
    override fun sendRawInput(data: ByteArray): Boolean {
        // Not implemented - use specific methods instead
        Log.w(TAG, "sendRawInput not implemented - use specific HID methods")
        return false
    }
}
